Parsimony/OptionSpec.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+
+namespace Parsimony
+{
+    /// <summary>
+    /// A specification for a CLI option.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of the option set.</typeparam>
+    /// <typeparam name="TValue">The type of this option's value.</typeparam>
+    public class OptionSpec<TOptions, TValue>
+        : OptionSpec<TOptions>
+        where TOptions : notnull, new()
+    {
+        private readonly Func<string, TValue> _parseFn;
+
+        private readonly Action<TOptions, TValue> _setFn;
+
+        /// <summary>
+        /// The default value for the option. If <see cref="OptionSpec{TOptions}.Required"/> is
+        /// <c>true</c> this value  will be ignored.
+        /// </summary>
+        public TValue Default { get; private set; }
+
+        /// <summary>
+        /// The <see cref="Type"/> of the option.
+        /// </summary>
+        public override Type OptionType => typeof(TValue);
+
+        // TODO: Create constructors for valid configurations?
+
+        /// <summary>
+        /// Creates a new <see cref="OptionSpec{TOptions,TValue}"/>.
+        /// </summary>
+        /// <param name="shortName">The short name of the option.</param>
+        /// <param name="longName">The long name of the option.</param>
+        /// <param name="required">Indicates whether the option is required.</param>
+        /// <param name="default">
+        /// The default value for the option (if 'required' is false).
+        /// </param>
+        /// <param name="parseFn">The function to use when parsing the option value.</param>
+        /// <param name="setFn">
+        /// The procedure to use to assign the option value to the <typeparamref name="TOptions"/>.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="parseFn"/> or <paramref name="setFn"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// Both <paramref name="shortName"/> and <paramref name="longName"/> are <c>null</c>.
+        /// </exception>
+        public OptionSpec(
+            char? shortName,
+            string? longName,
+            bool required,
+            TValue @default,
+            Func<string, TValue> parseFn,
+            Action<TOptions, TValue> setFn)
+            : base(shortName, longName, required)
+        {
+            Default = @default;
+            _parseFn = parseFn ?? throw new ArgumentNullException(nameof(parseFn));
+            _setFn = setFn ?? throw new ArgumentNullException(nameof(setFn));
+        }
+
+        /// <summary>
+        /// The function to use when parsing the option value.
+        /// </summary>
+        public TValue Parse(string input) => _parseFn(input);
+
+        /// <summary>
+        /// The procedure to use to assign the option's value to the option set.
+        /// </summary>
+        public void Set(TOptions options, TValue value) => _setFn(options, value);
```

Comment:
Why not just a readonly Selector type of property?

Reply:
not sure what you mean

Reply:
This was before the comment above but the idea being that it's odd to maintain an internal setter function and then reflect off the value from the call site if you're not using any internal state other than the function itself. It's a bit of a nit I just don't really like this pattern + the reflectybits a ton

Reply:
im not a huge fan either, but i was more concerned with supporting type safety in the context of constructing option specs and the future parser builder. going with the optoins-as-their-own-parsers idea will probably help with that